src/Build/BackEnd/Components/RequestBuilder/TargetBuilder.cs

```diff
@@ -766,5 +767,45 @@ private async Task<bool> CompleteOutstandingActiveRequests(string targetName)
 
             return false;
         }
+
+        private void ComputeAfterTargetFailures(string[] targetNames)
+        {
+            // Cache intermediate results so we don't visit a target more than once.
+            Dictionary<string, bool> resultsWithAfterTargets = new Dictionary<string, bool>();
+
+            foreach (string targetName in targetNames)
+            {
+                if (_buildResult.ResultsByTarget.ContainsKey(targetName))
+                {
+                    // Queue of targets waiting to be processed, seeded with the specific target for which we're computing AfterTargetsHaveFailed.
+                    Queue<string> targetsToCheckForAfterTargets = new Queue<string>();
+                    targetsToCheckForAfterTargets.Enqueue(targetName);
+
+                    while (targetsToCheckForAfterTargets.Count > 0)
+                    {
+                        string targetToCheck = targetsToCheckForAfterTargets.Dequeue();
+                        IList<TargetSpecification> targetsWhichRunAfter = _requestEntry.RequestConfiguration.Project.GetTargetsWhichRunAfter(targetToCheck);
+
+                        foreach (TargetSpecification afterTarget in targetsWhichRunAfter)
+                        {
+                            if (!resultsWithAfterTargets.TryGetValue(afterTarget.TargetName, out bool afterTargetFailed))
+                            {
+                                _buildResult.ResultsByTarget.TryGetValue(afterTarget.TargetName, out TargetResult result);
```

Comment:
minor: is there any value to assessing the outcome of `_buildResult.ResultsByTarget.TryGetValue(` ?

Reply:
Yes, the outcome is assessed on the following line with `result != null`. We want to make this defensive check anyway and can take advantage of the `TryGetValue` contract which says that if the key is not found, the "default value for the type of the value parameter" is returned. TValue is a reference type so the its default value is `null`. Two birds with one stone, if you will.